File: 0278-first-bad-version/0278-first-bad-version.cpp
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        int l=1;
        int r=n;
        int ans=-1;
        while(true){
            int m=l+(r-l)/2;
            if(!isBadVersion(m)){
                l=m+1;
            }
            else {
                if(isBadVersion(m-1)){
                    r=m-1;
                    ans=m-1;
                }
                else{
                    return m;
                }
            }
        }
        return ans;

    }
};

File: 0278-first-bad-version/0278-first-bad-version.cs
/* The isBadVersion API is defined in the parent class VersionControl.
      bool IsBadVersion(int version); */

public class Solution : VersionControl {
    public int FirstBadVersion(int n) {
        int l=0;
        int r=n;
        int ans=-1;
        while(true){
            int m=l+(r-l)/2;
            if(!IsBadVersion(m)){
                l=m+1;
            }
            else if(IsBadVersion(m-1)){
                r=m-1;
                ans=m-1;
            }
            else{
                return m;
            }
        }
        return ans;
    }
}